### PR TITLE
Suite refacto lightningcli

### PR DESCRIPTION
**Pull request by clemoule**
integration of mlflow_logger in lightningmodule
Modify configure_optimizer in lightningmodule
integration scheduler and optimizer args in config.yaml
rm use_lr_scheduler
rm no_log
modify test.yaml (for CI)
rm test_lightning_fit_inference from test_model.py
integrate ckpt loading and saving

### CODE REVIEW

File: py4cast/lightning.py [hunk 1]
```diff
@@ -163,12 +165,16 @@ def __init__(
...
-        use_lr_scheduler: bool = False,
-        no_log: bool = False,
         channels_last: bool = False,
         num_samples_to_plot: int = 1,
+        optimizer: str = "Adam", # name of the optimizer (adam, SGD)
+        weight_decay: float = 0.001, # deepNN : 0.0001 à 0.1, linear regression : 0.001 à 0.1, classification: 0.01 à 0.1
+        lr_scheduler: str = "cosine_scheduler", # name of the scheduler (StepLR, OneCycleLR, cosine_scheduler)
+        #useful if StepLR is used
+        StepLR_step_size: int = 10, # LR=LR*gamma every num_epoch=step_size  (usually 5-20, must divide max_epoch)
+        StepLR_gamma: float = 0.1,
```
Comment (LBerth) [inline py4cast/lightning.py:177]: variable names should contain only lower case

File: config/CLI/trainer.yaml [hunk 2]
```diff
@@ -46,6 +72,14 @@ trainer:  # Native training arguments of ligthining.pytorch.Trainer
   strategy: auto               # Training strategy alias (auto, ddp, fsdp, etc)
   accelerator: auto            # Type of device (auto, gpu, cpu)
   devices: auto                # Number of devices (auto or NUM)
-  limit_train_batches: null     # Limit num of batches per epoch (to do very short training)
-  limit_val_batches: null
-  limit_test_batches: null
\ No newline at end of file
+  limit_train_batches: 3     # Limit num of batches per epoch (to do very short training)
+  limit_val_batches: 3
+  limit_test_batches: 3
```
Comment (LBerth) [inline config/CLI/trainer.yaml:77]: put back to null

File: py4cast/lightning.py [hunk 3]
```diff
@@ -161,14 +159,17 @@ def __init__(
...
-        use_lr_scheduler: bool = False,
-        no_log: bool = False,
         channels_last: bool = False,
         num_samples_to_plot: int = 1,
+        optimizer_name: str = "AdamW",
+        lr: float = 0.001,
+        lr_scheduler: str = "cosine",
+        warmup_epochs: int = 1,
+        max_epochs: int = 50,
+        eta_min: int = 0,
```
Comment (LBerth) [inline py4cast/lightning.py:172]: rm from lightniing arguments please

File: py4cast/lightning.py [hunk 4]
```diff
@@ -583,6 +577,10 @@ def _common_step(
 
     def on_train_start(self):
         self.train_plotters = []
+        # print("Optimizer:")
+        # print(self.optimizers())
+        # print("Scheduler:")
+        # print(self.lr_schedulers())
```
Comment (LBerth) [inline py4cast/lightning.py:583]: rm prints please

File: tests/test_models.py [hunk 5]
```diff
@@ -264,3 +163,6 @@ def test_model_registry():
         "GraphLAM",
         "HiLAMParallel",
     }
+    print()
+    print("test_model_registry WORKS !")
+    print()
```
Comment (LBerth) [inline tests/test_models.py:168]: rm prints

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -52,5 +52,9 @@ jobs:
         coverage run -p bin/inference.py --dataset dummy --model_path /home/runner/work/py4cast/py4cast/logs/camp0/dummy/HalfUNet/runn_run_0/
         coverage run -p bin/train.py --precision 32 --model HiLAM --dataset dummy --epochs 1 --batch_size 1 --num_pred_steps_train 1 --limit_train_batches 1 --num_workers 1 --pin_memory
         coverage run -p bin/train.py --precision 32 --model UNETRPP --dataset dummy --epochs 1 --batch_size 1 --num_pred_steps_train 1 --limit_train_batches 1 --num_workers 1
+        coverage run -p bin/main.py fit --config config/CLI/trainer.yaml --config config/CLI/dataset/dummy.yaml --config config/CLI/model/unetrpp.yaml --trainer.fast_dev_run True
+        coverage run -p bin/main.py fit --config config/CLI/trainer.yaml --config config/CLI/dataset/dummy.yaml --config config/CLI/model/unetrpp.yaml --trainer.fast_dev_run False --trainer.max_epochs 1 --data.batch_size 1 --trainer.limit_train_batches 1
+
```
Comment (LBerth) [inline .github/workflows/tests.yml:57]: please add a test using lightning CLI "test" method

**Review by LBerth — APPROVED** [review]
good job !

File: lightning_logs/test
Comment (cbovalo) [inline lightning_logs/test:1]: Should this repo be committed ?

**Review by cbovalo — APPROVED** [review]

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -50,7 +50,12 @@ jobs:
...
+        coverage run -p bin/main.py fit --config config/CLI/trainer.yaml --config config/CLI/dataset/dummy.yaml --config config/CLI/model/unetrpp.yaml --trainer.fast_dev_run True
+        coverage run -p bin/main.py fit --config config/CLI/trainer.yaml --config config/CLI/dataset/dummy.yaml --config config/CLI/model/unetrpp.yaml --trainer.fast_dev_run False --trainer.max_epochs 1 --data.batch_size 1 --trainer.limit_train_batches 1 --trainer.logger.init_args.version version_test
+        coverage run -p bin/main.py fit --config config/CLI/trainer.yaml --config config/CLI/dataset/dummy.yaml --config config/CLI/model/unetrpp.yaml --trainer.fast_dev_run False --trainer.max_epochs 1 --data.batch_size 1 --trainer.limit_train_batches 1 --ckpt_path /home/runner/work/py4cast/py4cast/lightning_logs/version_test/checkpoints/last.ckpt
+        coverage run -p bin/main.py test --config config/CLI/trainer.yaml --config config/CLI/dataset/dummy.yaml --config config/CLI/model/unetrpp.yaml --trainer.fast_dev_run False --trainer.max_epochs 1 --data.batch_size 1 --trainer.limit_train_batches 1 --ckpt_path /home/runner/work/py4cast/py4cast/lightning_logs/version_test/checkpoints/last.ckpt
```
Comment (tourniert) [inline .github/workflows/tests.yml:57]: Why did you set `trainer.fast_dev_run` to `False` and not to `True` if then you set `--trainer.max_epochs 1 --trainer.limit_train_batches 1` ?
It looks the same in the doc: https://pytorch-lightning.readthedocs.io/en/1.0.8/debugging.html#fast-dev-run

File: bin/train.py [hunk 8]
```diff
@@ -275,45 +260,41 @@
...
+    dirpath=save_path,
+    filename="{epoch:02d}-{val_mean_loss:.2f}",  # Custom filename pattern
+    monitor="val_mean_loss",
+    mode="min",
+    save_top_k=1,  # Save only the best model
+    save_last=True,  # Also save the last model
+)
+callback_list.append(checkpoint_callback)
+callback_list.append(LearningRateMonitor(logging_interval="step"))
+callback_list.append(EarlyStopping(monitor="val_mean_loss", mode="min", patience=50))
```
Comment (tourniert) [inline bin/train.py:297]: A patience set to 50 looks a bit to much, may be set it to 5 ?

File: py4cast/lightning.py [hunk 9]
```diff
@@ -288,13 +280,27 @@ def __init__(
...
+            max_pred_step = self.num_pred_steps_val_test - 1
             self.rmse_psd_plot_metric = MetricPSDVar(pred_step=max_pred_step)
             self.psd_plot_metric = MetricPSDK(self.save_path, pred_step=max_pred_step)
             self.acc_metric = MetricACC(self.dataset_info)
+            print()
+            print(f"LOG DIR : {self.save_path}")
+            print()
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=0.001)
```
Comment (tourniert) [inline py4cast/lightning.py:295]: Don't you use the optimizer set in the CLI ?

Comment (clemoule) [reply py4cast/lightning.py:295]: to pass the test_model.py which needs a configure optimizer to proceed.

File: .github/workflows/tests.yml (hunk 7 above)
Comment (clemoule) [reply .github/workflows/tests.yml:57]: to detect (if they exists) loopholes and issues with logs, path, dirs that could arise with a proper fit (=! fit w/ fast_dev_run=true)

**Review by LBerth — APPROVED** [review]
